text/0813-template-layering.md

```diff
@@ -0,0 +1,984 @@
+so that their code can be agonistic against where the compilation is happening.
+However, since the intent is that build-time compilation should be _possible_,
+they should also take care to ensure they stay within the set of idiomatic set
+of usages to guarantee that is the case, and this re-export from the runtime
+module offers a way to signal that the deviations are deliberate.
+
+This RFC does not propose to add re-exports for other features found in the
+`@ember/template-compilation` package. Conceptually, it makes sense to provide
+the same ex-export for at least the `compile()` function. However, as we move
+to a world where components are used everywhere and the need for "standalone
+templates" dwindles, it is unclear that the `compile()` function is needed in
+the long run. In any case, if that turns out to be necessary, a future RFC can
+propose to add support for that and the precedence set forth in this RFC should
+make that a straightforward process.
+
+### TypeScript
+
+We don't expect this feature to pose any significant challenges in terms of
+TypeScript support, other than having a few overloaded signatures/purpose on
+the same `template()` function may make things a little cumbersome, but not
+prohibitively so. The signature of the function can be described as such:
+
+```ts
+// These types are defined in other RFCs
+interface Signature {
+  /* ... */
+}
+interface Component<S extends Signature> {
+  /* ... */
+}
+
+type Scope = () => Record<string, unknown>;
+
+function template<S extends Signature>(source: string): Component<S>;
+function template<S extends Signature>(
+  source: string,
+  scope: Scope
+): Component<S>;
+function template<S extends Signature, C extends Component<S>>(
+  source: string,
+  scope: Scope,
+  target: C
+): C;
+
+// For the build-time tagged template literal extension. By not specifying
+// additional arguments here, TypeScript actually enforces that the tagged
+// string literal cannot have any interpolation expressions.
+function template<S extends Signature>(
+  source: TemplateStringsArray
+): Component<S>;
+```
+
+One nice thing about this design is that there is a natural place for the type
+(generic) argument for the template's signature, negating the need for the
+special `<template[Signature]>` syntax.
+
+## How we teach this
+
+The teaching recommendations in [RFC #779][rfc-779] shall remain in full force.
+We continue to recommend teaching materials to be focused around `<template>`
+and `.gjs`. The `template()` feature is recommended only in standard JavaScript
+environments where adoption of `<template>` and `.gjs` is not feasible.
+
+Because `template()` is a standard JavaScript API that has a "real" import
+location, there is a natural place to document and describe its behavior in the
+API docs. This should be the primary reference of the feature, and should also
+where we document the idiomatic subset as well.
```

Comment:
We should use that as a hook for describing the whole end-to-end bits in the API docs, including `<template>`, too, since as you note below it's the actual desugaring for that.